In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.decomposition import PCA
from sklearn.svm import SVR
from sklearn.metrics import r2_score, mean_squared_error
pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_csv('data/07_flats_EDA.csv')
df.dropna(inplace=True)

In [3]:
df['price_cr'] = df['price_cr']*1.7    #since, the dataset is 4 years old, increasing prices by 70%

In [4]:
df.head(2)

,area_sqm,bedroom,bathroom,balcony,addl_room,sector,landmark_count,has_hospital_or_clinic,has_police_stn,has_fire_stn,has_mall,has_market,has_park,has_bank_or_atm,has_institute,has_airport,has_rly_stn,has_metro_stn,has_expy,has_petrol_pump,has_hotel,has_temple_or_church,has_famous_chowk,has_club,has_stadium,has_restro_cafe,area_type,floor_rise,age,society,city,price_cr
0,83.61,2.0,2.0,1.0,0.0,sector 7,50.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,carpet,mid,new,maa bhagwati residency,gurgaon,0.9945
1,60.39,2.0,2.0,1.0,0.0,sector 3,48.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,carpet,low,old,apna enclave,gurgaon,1.1050


In [5]:
le = LabelEncoder()
for col in df.columns:
    if df[col].dtype == 'object':
        df[col] = le.fit_transform(df[col])

In [6]:
X = df.iloc[:,1:]
y = df['price_cr']

In [7]:
#train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
#Scaling X
sc = StandardScaler()
X_train_scaled = sc.fit_transform(X_train)
X_test_scaled = sc.transform(X_test)

In [9]:
X_train_scaled.shape[1]

31

In [10]:
X_train_scaled = pd.DataFrame(X_train_scaled)

In [11]:
#using PCA to reduce dimensionality
pca = PCA(n_components=0.95)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

In [12]:
X_train_pca.shape[1]

26

In [13]:
pca.components_

array([[ 9.07822865e-02,  1.30748853e-01,  1.50463147e-01,
         1.33106650e-01,  2.58252437e-02, -2.98749748e-01,
         1.34969338e-02, -4.63363766e-02, -1.69241351e-02,
         2.96018064e-01, -1.49700794e-01,  1.41276081e-01,
        -3.16383338e-01,  2.63714885e-01,  3.18023746e-01,
         4.35158053e-02,  2.84915499e-02,  2.23232728e-01,
        -2.43529534e-01,  2.26599429e-01, -2.98317056e-01,
        -1.12403803e-01, -1.19547151e-02,  1.14837431e-01,
        -2.82327904e-01,  1.65964383e-01, -9.80217449e-02,
        -1.87345778e-01, -1.91676417e-03, -5.23260212e-02,
         9.02315388e-02],
       [ 4.19788838e-01,  4.33388036e-01,  2.87559324e-01,
         2.75080694e-01, -6.32039922e-02,  2.14248742e-01,
         1.60768844e-01,  5.80818524e-02,  1.82867075e-02,
        -1.25634882e-02,  1.57165147e-01,  1.69183133e-01,
         1.73224702e-01,  2.28641330e-04,  9.68951576e-03,
        -3.64875126e-02,  1.46405446e-01, -1.03625027e-01,
         4.93025325e-02,  6.23

In [14]:
pca.explained_variance_

array([5.03670091, 3.18792607, 2.01097724, 1.73579498, 1.40030085,
       1.37400313, 1.21221471, 1.1017453 , 1.03006511, 1.01331613,
       0.98954204, 0.9071451 , 0.89007557, 0.78953632, 0.77139363,
       0.72773651, 0.69485352, 0.66471865, 0.62136946, 0.60363811,
       0.56355481, 0.54730291, 0.49626062, 0.45436567, 0.44325493,
       0.42146521])

In [15]:
pca.explained_variance_ratio_

array([0.16240428, 0.10279206, 0.06484231, 0.05596928, 0.04515155,
       0.0443036 , 0.03908687, 0.03552487, 0.0332136 , 0.03267355,
       0.03190697, 0.02925015, 0.02869976, 0.02545795, 0.02487295,
       0.02346527, 0.02240498, 0.02143331, 0.02003555, 0.01946381,
       0.01817136, 0.01764733, 0.01600152, 0.01465065, 0.01429239,
       0.0135898 ])

In [16]:
pca.explained_variance_ratio_.sum()

0.9573057065612774

In [17]:
y_train.unique()

array([ 0.9945,  7.4035,  2.431 ,  1.989 ,  3.7128,  4.0885,  5.4145,
        1.0387,  3.2487,  2.9614,  1.1713,  3.0056,  3.978 ,  7.735 ,
        3.315 ,  2.9172,  2.5636,  6.6079,  2.3868,  0.5525,  3.094 ,
        2.1879,  0.9282, 18.2325,  1.105 ,  3.4255,  6.409 ,  2.7625,
       14.365 ,  1.326 , 11.05  ,  2.3205,  4.2211,  1.1934,  5.7902,
        0.663 ,  1.768 ,  0.4862,  0.8398,  3.536 ,  3.2045,  1.9227,
        2.21  ,  2.4752,  2.652 ,  3.757 ,  6.188 ,  9.061 ,  2.9835,
        3.1382, 10.4975,  0.884 ,  3.8675,  3.8896,  1.6575,  1.0608,
        2.6078,  4.199 ,  6.5195,  2.5415,  8.177 ,  3.8233,  2.1658,
        8.84  ,  4.0443, 10.9174,  1.9448,  7.2709,  2.873 ,  4.1327,
        8.0665,  2.0995,  0.6188,  4.0222, 17.238 ,  5.083 ,  0.5746,
        3.0719,  2.8951,  3.6686,  4.5305,  1.8785,  4.7957,  8.4643,
        1.9669,  2.7846,  3.2929,  4.4863,  0.8619,  1.547 ,  5.1935,
       19.89  ,  7.956 ,  1.0166,  3.6465,  3.9338,  5.8565,  0.5967,
        1.7238,  1.8

In [18]:
#model 1: RF Regression
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)

print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2_score(y_test, y_pred)}")

RMSE     : 0.9556297425090575
R2_score : 0.8908913867116075


In [19]:
#model 2: XG Boost Regression
from xgboost import XGBRegressor
model = XGBRegressor(objective='reg:squarederror',n_estimators=100, 
                         learning_rate=0.1,max_depth=5,random_state=42)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)

print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2_score(y_test, y_pred)}")

RMSE     : 0.9718142283755611
R2_score : 0.887164378152375


In [20]:
#model 3: SVR Regression
from sklearn.svm import SVR
model = SVR(kernel='rbf', C=100, gamma=0.1, epsilon=.1)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)

print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2_score(y_test, y_pred)}")

RMSE     : 0.9241702283699393
R2_score : 0.8979568943415293


In [27]:
#model 4: SVR Regression
from sklearn.svm import SVR
model = SVR(kernel='rbf', C=70, gamma=0.06, epsilon=.14)
model.fit(X_train_pca, y_train)
y_pred = model.predict(X_test_pca)
r2 = r2_score(y_test, y_pred)
err = 1 - r2
lb = 1 - err
ub = 1 + err
print(f"RMSE     : {np.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"R2_score : {r2}")

RMSE     : 0.5863055248667177
R2_score : 0.958929717194295


#### the best model is Support Vector Regressor (SVR) with 86% R2 Score